src/app/customizable-demo-form.component.ts

```diff
@@ -164,6 +173,73 @@ export class CustomizableDemoForm {
   // Whether to hide the loading icon when the score is below the minimum
   // threshold to show feedback.
   hideLoadingIconForScoresBelowMinThreshold = false;
+  // Whether to always hide the loading icon.
+  alwaysHideLoadingIcon = false;
+
+  demoSettings: DemoSettings|null = null;
+
+  constructor(private route: ActivatedRoute, private router: Router) {}
+
+  ngOnInit() {
+    this.route.params.subscribe((params: Params) => {
+      if (params['uiSettings']) {
+        const decodedUISettings: UISettings = JSON.parse(
```

Comment:
These seem like they'd be better passed in as optional url parameters rather than path components, what's the reasoning for doing it like this?

Reply:
Thinking out loud...

These are really only optional for the initial load of the page, all subsequent init calls would come from a saved url or from manual navigation below, so we'd expect these settings to be present in some way. So I'm not sure optional is the best way to describe them.

That being said, there are multiple ways to do this (pass these url parameters) that are valid: Angular does provide the means of accessing the same url with different numbers of parameters which I have done here (having one route with just customize/ and another with customize/:uiSettings/:demoSettings.

But there are also these docs [https://angular.io/guide/router#heroes-list-optionally-selecting-a-hero], which offer a way of doing it with optional parameters the "angular way", which looks like the only difference is the style of the url and that I pass an object instead of an array when calling navigate().

The docs don't make it clear what the benefit of doing it one way over the other is, so I guess the long answer is I'm really not sure? I think either way this code wouldn't change.

Reply:
Okay, I'm fine with that but having json-encoded URL paths containing config variables feels a bit odd from the user stadpoint.

Reply:
I understand! To be fair this is not for the user but for us to use to make saving URLs easier when doing user studies so the urls are not intended to be human readable.